libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -402,36 +402,50 @@ - (NSString *)sfHybridViewUserAgentString
     return userAgentString;
 }
 
-- (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrl returnUrlIsEncoded:(BOOL)isEncoded createAbsUrl:(BOOL)createAbsUrl
+- (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrlString returnUrlIsEncoded:(BOOL)isEncoded createAbsUrl:(BOOL)createAbsUrl
 {
+    // Special case: if returnUrlString itself is a frontdoor.jsp URL, parse its parameters and rebuild.
+    if ([returnUrlString containsString:@"frontdoor.jsp"]) {
```

Comment:
Moved this to the top, and farmed it out to a helper method, since it's a self-contained code block that otherwise doesn't apply here.